tests/staggered_gsmear_test_utils.h

```diff
@@ -0,0 +1,405 @@
+#pragma once
+
+#include <iostream>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <algorithm>
+
+#include <quda.h>
+#include <gauge_field.h>
+#include <dirac_quda.h>
+#include <misc.h>
+#include <host_utils.h>
+#include <command_line_params.h>
+
+#include <gsmear_test_helpers.h>
+#include <assert.h>
+#include <gtest/gtest.h>
+
+using namespace quda;
+
+void initExtendedField(void* sitelink_ex[4], void* sitelink[4]) 
+{	
+  int X1=Z[0];
+  int X2=Z[1];
+  int X3=Z[2];
+  int X4=Z[3];
+
+  for(int i=0; i < V_ex; i++){
+    int sid = i;
+    int oddBit=0;
+    if(i >= Vh_ex){
+      sid = i - Vh_ex;
+      oddBit = 1;
+    }
+
+    int za = sid/E1h;
+    int x1h = sid - za*E1h;
+    int zb = za/E2;
+    int x2 = za - zb*E2;
+    int x4 = zb/E3;
+    int x3 = zb - x4*E3;
+    int x1odd = (x2 + x3 + x4 + oddBit) & 1;
+    int x1 = 2*x1h + x1odd;
+
+
+    if( x1< 2 || x1 >= X1 +2
+        || x2< 2 || x2 >= X2 +2
+        || x3< 2 || x3 >= X3 +2
+        || x4< 2 || x4 >= X4 +2){
+
+      continue;
+
+    }
+
+    x1 = (x1 - 2 + X1) % X1;
+    x2 = (x2 - 2 + X2) % X2;
+    x3 = (x3 - 2 + X3) % X3;
+    x4 = (x4 - 2 + X4) % X4;
+
+    int idx = (x4*X3*X2*X1+x3*X2*X1+x2*X1+x1)>>1;
+    if(oddBit){
+      idx += Vh;
+    }
+    for(int dir= 0; dir < 4; dir++){
+      char* src = (char*)sitelink[dir];
+      char* dst = (char*)sitelink_ex[dir];
+      memcpy(dst + i * gauge_site_size * host_gauge_data_type_size,
+             src + idx * gauge_site_size * host_gauge_data_type_size, gauge_site_size * host_gauge_data_type_size);
+    }//dir
+  }//i
+  return;
+}
+
+
+gsmear_test_type gtest_type = gsmear_test_type::TwoLink;
+
+CLI::TransformPairs<gsmear_test_type> gtest_type_map {
+  {"TwoLink", gsmear_test_type::TwoLink}, {"GaussianSmear", gsmear_test_type::GaussianSmear}
+};
+
+struct GSmearTime {//DslashTime
+  double event_time;
+  double cpu_time;
+  double cpu_min;
+  double cpu_max;
+
+  GSmearTime() : event_time(0.0), cpu_time(0.0), cpu_min(DBL_MAX), cpu_max(0.0) { }
+};
+
+struct StaggeredGSmearTestWrapper {//
+
+  bool is_ctest = false; 
+
+  // Allocate host staggered gauge fields
+  void* qdp_inlink[4]    = {nullptr,nullptr,nullptr,nullptr};
+  void* qdp_twolnk[4]    = {nullptr,nullptr,nullptr,nullptr};
+  void* qdp_ref_twolnk[4]= {nullptr,nullptr,nullptr,nullptr};     
+  void* qdp_inlink_ex[4] = {nullptr,nullptr,nullptr,nullptr};//sitelink_ex
+  
+  void *milc_inlink       = nullptr;
+  void *milc_twolnk       = nullptr;
+  //
+  GaugeField *cpuTwoLink  = nullptr;
+
+  QudaGaugeParam  gauge_param;//
+  QudaInvertParam inv_param;
+  //
+  std::unique_ptr<ColorSpinorField> spinor;
```

Comment:
We never need to use smart pointers now for `ColorSpinorField` objects now that we have both default and move constructors.  These should be replaced with simply `ColorSpinorField` objects.

Reply:
not sure since it is in fact a conditional instantiation (e.g., the object is not needed for pure two-link tests)

Reply:
That doesn't matter, since `ColorSpinorField` has a default constructor, so it's easy enough to conditionally create the object.

Reply:
done